File: node_project_test.py
import unittest
import node_project
import os
import shutil
import json

class NodeProjectTest(unittest.TestCase):
    
    current_dir = None
    test_directory = "test"
    
    def setUp(self):
        self.current_dir = os.getcwd()
        os.mkdir(self.test_directory)
        os.chdir(self.test_directory)
        
    def tearDown(self):
        os.chdir(self.current_dir)
        shutil.rmtree(self.test_directory)

    @unittest.skip("takes too long")
    def testAddDevDependencies(self):
        node_project.addDevDependencies()
        
        self.assertTrue(os.path.exists("node_modules"))
        self.assertTrue(os.path.exists("package.json"))
            
    def testAddScripts(self):
        test_object = {"test":"test"}
        with open('package.json', 'w') as package_json:
            json.dump(test_object, package_json)
        
        node_project.addScripts()

        with open('package.json', 'r') as package_json:
            result = json.load(package_json)
            keys = result.keys()
            self.assertTrue("test" in keys)
            self.assertTrue("scripts" in keys)
            self.assertTrue("integration" in result["scripts"])

    def testSetupTsConfig(self):
        node_project.setupTsConfig()

        with open('tsconfig.json', 'r') as tsconfig_json:
            result = json.load(tsconfig_json)
            keys = result.keys()
            self.assertTrue("compilerOptions" in keys)
            self.assertTrue("exclude" in keys)
            
    def testSetupGitIgnore(self):
        node_project.setupGitIgnore()

        with open('.gitignore', 'r') as gitignore:
            result = gitignore.read()

            self.assertNotEqual(-1, result.find("node_modules"))
            self.assertNotEqual(-1, result.find("build"))
            
        

File: node_project/nodeProject_test.go
package main

import (
	"encoding/json"
	"os"
	"strings"
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestInitWorkingDirectory(t *testing.T) {

	testDir := "testDirectory"

	initWorkingDirectory(testDir)
	cwd, _ := os.Getwd()
	assert.Contains(t, cwd, testDir)
	destroyTestDirectory()
}

func TestAddDevDependencies(t *testing.T) {

	t.Skip("Takes too long")

	setupTestDirectory()
	addDevDependencies()

	dat, _ := os.ReadFile("package.json")

	assert.True(t, strings.Contains(string(dat), "jest"))
	destroyTestDirectory()
}

func TestAddScripts(t *testing.T) {

	setupTestDirectory()

	dummyMap := map[string]string{"dummy": "dummy"}

	marshaledDummy, _ := json.Marshal(dummyMap)

	os.WriteFile("package.json", marshaledDummy, 0644)

	addScripts()

	dat, _ := os.ReadFile("package.json")

	var packageJson map[string]interface{}

	json.Unmarshal(dat, &packageJson)

	assert.NotNil(t, packageJson["scripts"])
	scriptsJson := (packageJson["scripts"]).(map[string]interface{})
	assert.NotNil(t, scriptsJson["build"].(string))
	assert.True(t, strings.Contains(string(dat), "prod"))

	destroyTestDirectory()
}

func TestSetupTsCOnfig(t *testing.T) {
	setupTestDirectory()

	setupTsConfig()

	dat, _ := os.ReadFile("tsconfig.json")

	assert.True(t, strings.Contains(string(dat), "build/src"))

	destroyTestDirectory()
}

func TestSetupGitgnore(t *testing.T) {
	setupTestDirectory()

	setupGitIgnore()

	dat, _ := os.ReadFile(".gitignore")

	assert.True(t, strings.Contains(string(dat), "node_modules"))

	destroyTestDirectory()
}

func TestSetupJestConfig(t *testing.T) {
	setupTestDirectory()

	setupJestConfig()

	dat, _ := os.ReadFile("jest.config.js")

	assert.True(t, strings.Contains(string(dat), "preset"))

	destroyTestDirectory()
}

func setupTestDirectory() {
	testDirectory := "testDirectory"
	os.Mkdir(testDirectory, 0755)
	os.Chdir(testDirectory)
}

func destroyTestDirectory() {
	testDirectory := "testDirectory"
	os.Chdir("..")
	os.RemoveAll(testDirectory)
}
